bazel/p4_library.bzl

```diff
@@ -120,3 +120,103 @@ p4_library = rule(
     incompatible_use_toolchain_transition = True,
     toolchains = ["@bazel_tools//tools/cpp:toolchain_type"],
 )
+
+def _p4_graphs_impl(ctx):
+    p4c = ctx.executable._p4c
+    p4file = ctx.file.src
+    p4deps = ctx.files._p4include + ctx.files.deps
+    output_file = ctx.outputs.out_file
+    graph_dir = output_file.path + "-graphs-dir"
+    args = [
+        p4file.path,
+        "--std",
+        ctx.attr.std,
+        "--arch",
+        ctx.attr.arch,
+        "--graphs-dir",
+        graph_dir,
+    ]
+    if ctx.attr.extra_args:
+        args.append(ctx.attr.extra_args)
+
+    include_dirs = {d.dirname: 0 for d in p4deps}  # Use dict to express set.
+    include_dirs["."] = 0  # Enable include paths relative to workspace root.
+    args += [("-I" + dir) for dir in include_dirs.keys()]
+
+    cpp_toolchain = find_cpp_toolchain(ctx)
+    ctx.actions.run_shell(
+        command = """
+            # p4c invokes cc for preprocessing; we provide it below.
+            function cc () {{ "{cc}" "$@"; }}
+            export -f cc
+            # Create the output directory
+            mkdir "{graph_dir}"
+            # Run the compiler
+            "{p4c}" {p4c_args}
+            # Merge all output graphs
+            cat "{graph_dir}"/* > {output_file}
```

Comment:
Nit: "{graph_dir}"/* -> "{graph_dir}/*"
(unless this brakes the * somehow)

Reply:
That breaks the shell globbing by *, I added a comment to explain it.

Reply:
Thanks for adding the coment!